realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,269 @@
+            cache = new RealmCache(configuration);
+            // The new cache should be added to the map later
+            isCacheInMap = false;
+        } else {
+            // != takes no time, check it first
+            if (cache.configuration != configuration && cache.configuration.hashCode() != configuration.hashCode()) {
+                // throw the exception
+                validateAgainstExistingConfigurations(configuration, cache.configuration);
+            }
+        }
+
+        RefAndCount refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+
+        if (refAndCount.localRealm.get() == null) {
+            // Create a new local Realm instance
+            BaseRealm realm;
+
+            if (realmClass == Realm.class) {
+                // RealmMigrationNeededException might be thrown here
+                realm = Realm.createInstance(configuration, cache.typedColumnIndices);
+            } else {
+                realm = DynamicRealm.createInstance(configuration);
+            }
+            // The Realm instance has been created without exceptions. Cache and reference count can be updated now.
+
+            // The cache is not in the map yet. Add it to the map after the Realm instance created successfully.
+            if (!isCacheInMap) {
+                cachesMap.put(configuration.getPath(), cache);
+            }
+            refAndCount.localRealm.set(realm);
+            refAndCount.localCount.set(0);
+        }
+
+        Integer refCount = refAndCount.localCount.get();
+        if (refCount == 0) {
+            if (realmClass == Realm.class && refAndCount.globalCount == 0) {
+                cache.typedColumnIndices = refAndCount.localRealm.get().schema.columnIndices;
+            }
+            // This is the first instance in current thread, increase the global count.
+            refAndCount.globalCount++;
+        }
+        refAndCount.localCount.set(refCount + 1);
+
+        return refAndCount.localRealm.get();
+    }
+
+    /**
+     * Releases a given {@link Realm} or {@link DynamicRealm} from cache. The instance won't be closed by this function.
+     *
+     * @param realm Realm instance to be released from cache.
+     * @return local reference count to the type Realm after releasing. {@code -1} if the Realm instance is not in the
+     *         cache.
+     */
+    static synchronized int release(BaseRealm realm) {
+        String canonicalPath = realm.getPath();
+        RealmCache cache = cachesMap.get(canonicalPath);
+        Integer refCount = null;
+        RefAndCount refAndCount = null;
+
+        if (cache != null) {
+            refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realm.getClass()));
+            refCount = refAndCount.localCount.get();
+        }
+        if (refCount == null) {
+            refCount = 0;
+        }
+
+        if (refCount <= 0) {
+            RealmLog.w("Realm " + canonicalPath + " is not in the cache anymore.");
```

Comment:
More user friendly warning needed. Isn't this replacing the old "Calling Realm.close() too many times..." message?